server/src/main/java/org/candlepin/model/EntitlementCurator.java

```diff
@@ -573,60 +573,69 @@ private boolean poolOverlapsRange(Pool p, Date startDate, Date endDate) {
+                "     SET dirty = true" +
+                "     WHERE id IN (" +
+                "         SELECT * FROM (" +
+                "             SELECT distinct ent.id" +
+                "             FROM cp_entitlement ent" +
+                "             INNER JOIN cp_pool pool" +
+                "                 ON ent.pool_id = pool.id" +
+                "             INNER JOIN cp2_pool_provided_products providedPr" +
+                "                 ON pool.id = providedPr.pool_id" +
+                "             INNER JOIN cp2_products product" +
+                "                 ON providedPr.product_uuid = product.uuid" +
+                "             INNER JOIN cp2_product_content productCon" +
+                "                 ON product.uuid = productCon.product_uuid" +
+                "             INNER JOIN cp2_content content" +
+                "                 ON productCon.content_uuid = content.uuid" +
+                "             INNER JOIN cp2_content_modified_products modifiedPr" +
+                "                 ON content.uuid=modifiedPr.content_uuid" +
+                "             WHERE pool.endDate >= current_date" +
+                "               AND ent.owner_id = :owner" +
+                "               AND ent.dirty = false" +
+                "               AND (ent.id NOT IN (:ein))" +
+                "               AND (EXISTS (SELECT entIn.id" +
+                "                            FROM cp_entitlement entIn" +
+                "                            INNER JOIN cp_consumer consumerIn" +
+                "                                ON entIn.consumer_id = consumerIn.id" +
+                "                            INNER JOIN cp_pool poolIn" +
+                "                                ON entIn.pool_id = poolIn.id" +
+                "                            INNER JOIN cp2_products productIn" +
+                "                                ON poolIn.product_uuid = productIn.uuid" +
+                "                            LEFT OUTER JOIN cp2_pool_provided_products providedPrIn" +
+                "                                ON poolIn.id = providedPrIn.pool_id" +
+                "                            LEFT OUTER JOIN cp2_products productIn2" +
+                "                                ON providedPrIn.product_uuid = productIn2.uuid" +
+                "                            WHERE (entIn.id IN (:ein))" +
+                "                               AND consumerIn.id = ent.consumer_id" +
+                "                               AND poolIn.endDate >= pool.startDate" +
+                "                               AND poolIn.startDate <= pool.endDate" +
+                "                               AND (productIn2.product_id = modifiedPr.element " +
+                "                                 OR productIn.product_id = modifiedPr.element)))) as e)";
```

Comment:
Just to make sure I understand this correctly, I've written what I interpret the intent of this query to be in a sort of pseudo code. Could you verify? If there is anything incorrect in my interpretation, could you provide one?

```
- For each entitlement:
    - Get a set of modified products:
        - Get the entilement's pool
        - Get the pool's product and provided products (Note: NOT its derived products)
        - Get the product's content
        - Get the content's modified products
        - Fetch the modified products for entitlements with pools that haven't yet expired*,
          matching the given owner and with IDs NOT in the set of given entitlements

    - For each modified product:
        - Check for related entitlements
            - Get the entitlements used by the given consumer
            - Get each entitlement's pool
            - Get the pool's products and provided products
            - Fetch the products with IDs matching the modified product and overlapping pool durations
              and the entitlement ID is in the given set of entitlements

        - If there is a related entitlement, set the entitlement's dirty flag


    * This will also match future pools
```